source/extensions/common/tap/tap_config_base.h

```diff
@@ -18,9 +18,34 @@ class Utility {
 public:
   /**
    * Add body data to a tapped body message, taking into account the maximum bytes to buffer.
+   * @param max_buffered_bytes supplies the maximum bytes to store, if truncation occurs the
+   *        truncation flag will be set.
+   * @param data supplies the data to buffer.
+   * @param buffer_start_offset supplies the offset within data to start buffering.
+   * @param buffer_length_to_copy supplies the length of the data to buffer.
+   * @return whether the buffered data was truncated or not.
    */
-  static void addBufferToProtoBytes(envoy::data::tap::v2alpha::Body& output_body,
-                                    uint32_t max_buffered_bytes, const Buffer::Instance& data);
+  static bool addBufferToProtoBytes(envoy::data::tap::v2alpha::Body& output_body,
+                                    uint32_t max_buffered_bytes, const Buffer::Instance& data,
+                                    uint32_t buffer_start_offset, uint32_t buffer_length_to_copy);
+
+  /**
+   * Trim a container that contains buffer raw slices so that the slices start at an offset and
+   * only contain a specific length. No slices are removed from the container, but their length
+   * may be reduced to 0.
+   */
+  template <typename T> static void trimSlices(T& slices, uint32_t start_offset, uint32_t length) {
+    for (auto& slice : slices) {
```

Comment:
I think I'd find it easier to read if this trim logic was effectively inlines in the `mutable_as_bytes` append loop. Or, do you plan on using this helper elsewhere?

Reply:
Ultimately we should move this into the buffer API, and it's much easier to test this way. This type of logic gives me a huge headache so would rather keep it like this unless there is a large reason to change it. Will add a comment.